### PR TITLE
En 3693 interceptors for rewards

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: process/block/preprocess/rewardsHandler.go [hunk 1]
```diff
@@ -8,79 +8,182 @@ import (
...
+	return shardId, nil
+}
+
 // CreateAllInterMiniBlocks creates miniblocks from process transactions
 func (rtxh *rewardsHandler) CreateAllInterMiniBlocks() map[uint32]*block.MiniBlock {
-	calculatedRewardTxs := rtxh.CreateAllUTxs()
+	rtxh.mutGenRewardTxs.Lock()
+	calculatedRewardTxs := make([]data.TransactionHandler, 0)
+	rtxh.feeRewards = rtxh.createRewardFromFees()
+	rtxh.addTransactionsToPool(rtxh.feeRewards)
```
Comment (sasurobert) [inline process/block/preprocess/rewardsHandler.go]: if we go with the SCR way of processing there is not need to add the created transactions into pool.

File: process/block/preprocess/transactions_test.go [hunk 2]
```diff
@@ -4,6 +4,8 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
```
Comment (sasurobert) [inline process/block/preprocess/transactions_test.go]: goimports.

File: process/coordinator/process.go [hunk 3]
```diff
@@ -358,6 +377,21 @@ func (tc *transactionCoordinator) ProcessBlockTransaction(
 		}
 	}
 
+	// create the reward txs and make them available for processing
+	mbRewards := tc.createRewardsMiniBlocks()
```
Comment (sasurobert) [inline process/coordinator/process.go]: no need to copy data from intermediate processor to preprocessor.

File: process/coordinator/process.go [hunk 4]
```diff
@@ -489,10 +523,19 @@ func (tc *transactionCoordinator) CreateMbsAndProcessTransactionsFromMe(
...
 
-	rewardsMBs := tc.createRewardsMiniBlocks()
-	if len(rewardsMBs) > 0 {
-		miniBlocks = append(miniBlocks, rewardsMBs...)
+	rewardMb := tc.createRewardsMiniBlocks()
+	if len(rewardMb) == 0 {
+		log.Error("could not create reward mini-blocks")
+	}
+
+	rewardsPreProc := tc.getPreProcessor(block.RewardsBlock)
```
Comment (sasurobert) [inline process/coordinator/process.go]: when creating the reward miniblock, the intermediate processor could also update the state, process effectivele the rewards.

File: process/rewardTransaction/process.go [hunk 5]
```diff
@@ -64,32 +71,37 @@ func (rtp *rewardTxProcessor) ProcessRewardTransaction(rTx *rewardTx.RewardTx) e
 	if rTx == nil {
 		return process.ErrNilRewardTransaction
 	}
+	if rTx.Value == nil {
+		return process.ErrNilValueFromRewardTransaction
+	}
+
+	rtp.mutRewardsForwarder.Lock()
+	err := rtp.rewardTxForwarder.AddIntermediateTransactions([]data.TransactionHandler{rTx})
```
Comment (sasurobert) [inline process/rewardTransaction/process.go]: add only those transactions which are created by the current shard, the others can be simply processed.

File: process/interface.go [hunk 6]
```diff
@@ -29,6 +29,12 @@ type RewardTransactionProcessor interface {
 	IsInterfaceNil() bool
 }
 
+// RewardTransactionPreProcessor prepares the processing of reward transactions
+type RewardTransactionPreProcessor interface {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/interface.go]: done

File: process/coordinator/process.go (hunk 4 above)
Comment (AdoAdoAdo) [reply process/coordinator/process.go]: moved this away from tx coordinator

File: process/block/preprocess/rewardTxPreProcessor.go [hunk 7]
```diff
@@ -212,11 +209,36 @@ func (rtp *rewardTxPreprocessor) ProcessBlockTransactions(body block.Body, round
 	return nil
 }
 
+// AddComputedRewardMiniBlocks adds to the local cache the reward transactions from the given miniblocks
+func (rtp *rewardTxPreprocessor) AddComputedRewardMiniBlocks(computedRewardMiniblocks block.MiniBlockSlice) {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/block/preprocess/rewardTxPreProcessor.go]: preprocessor should launch processing of all transactions where current shard is destination, it does not matter which is the source shard, so it needs to be added here.

**Comment by AdoAdoAdo** [comment]
> We should start writing the code in a more structured manner. No new structs should be added without unit tests as subtle bugs (invisible to integration tests) can hide really well.

As discussed this is a work in progress and unit tests are coming in a separate pull request.

File: data/address/specialAddresses.go [hunk 8]
```diff
@@ -78,6 +83,14 @@ func (sp *specialAddresses) ConsensusRewardAddresses() []string {
 	return sp.consensusRewardAddresses
 }
 
+func (sp *specialAddresses) Round() uint64 {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply data/address/specialAddresses.go]: done

File: data/address/specialAddresses.go [hunk 9]
```diff
@@ -78,6 +83,14 @@ func (sp *specialAddresses) ConsensusRewardAddresses() []string {
...
+	return sp.round
+}
+
+func (sp *specialAddresses) Epoch() uint32 {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply data/address/specialAddresses.go]: done

File: process/block/preprocess/rewardsHandler.go [hunk 10]
```diff
@@ -191,26 +314,17 @@ func (rtxh *rewardsHandler) createCommunityTx() *rewardTx.RewardTx {
...
-	rewardsForConsensus := rtxh.createRewardTxsForConsensusGroup()
-
-	rewardTxs = append(rewardTxs, rewardsFromFees...)
-	rewardTxs = append(rewardTxs, rewardsForConsensus...)
-
-	return rewardTxs
-}
-
-func (rtxh *rewardsHandler) createRewardTxsFromFee() []data.TransactionHandler {
+// createRewardFromFees creates the reward transactions from accumulated fees
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/block/preprocess/rewardsHandler.go]: done

File: process/factory/shard/intermediateProcessorsContainerFactory.go [hunk 11]
```diff
@@ -49,6 +51,9 @@ func NewIntermediateProcessorsContainerFactory(
 	if store == nil || store.IsInterfaceNil() {
 		return nil, process.ErrNilStorage
 	}
+	if poolsHolder == nil {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/factory/shard/intermediateProcessorsContainerFactory.go]: done

**Pull request merged**